### PR TITLE
giga speech dataloader

### PR DESCRIPTION
**Pull request by glynpu**
Hi Dan, @danpovey 
Is this pr what you want about gigaspeech dataloader?

Example usage:
```
import argparse                                                                 
from gigaspeech_datamodule import GigaSpeechAsrDataModule                       
                                                                                
def get_parser():                                                               
    parser = argparse.ArgumentParser(                                           
        formatter_class=argparse.ArgumentDefaultsHelpFormatter                  
    )                                                                           
    return parser                                                               
                                                                                
parser = get_parser()                                                           
GigaSpeechAsrDataModule.add_arguments(parser)                                   
args = parser.parse_args()                                                      
gigaspeech = GigaSpeechAsrDataModule(args)                                      
train_dl = gigaspeech.train_dataloaders()                                       
valid_dl = gigaspeech.valid_dataloaders()                                       
batch=next(iter(train_dl))                                                      
print(batch["supervisions"]["text"][0:5])                                       
print(batch["inputs"].shape)                                                    
batch=next(iter(train_dl))                                                      
print(batch["supervisions"]["text"][0:5])                                       
print(batch["inputs"].shape)


```

Expected output:

```
['HERE ARE JUST A FEW OF THE SEEMINGLY UNREAL <COMMA> ALBEIT <COMMA> ENTIRELY TRUE HIGHLIGHTS FROM HIS RESUME <PERIOD>', "HE'S A FORMER MTV BREAK DANCER IN TAIWAN <COMMA>", 'THE ONE TIME NATIONAL CHINESE KICKBOXING CHAMPION <COMMA>', 'THE FIRST AMERICAN TO HOLD THE GUINNESS WORLD RECORD IN TANGO <COMMA> A HORSEBACK ARCHER <COMMA> PRINCETON UNIVERSITY GUEST LECTURER <COMMA> AND ANGEL INVESTOR WHO HAS RACKED UP A SERIOUS STRING OF ENTREPRENEURIAL MAGA HITS <COMMA>', 'INCLUDING UBER <COMMA> FACEBOOK <COMMA> TWITTER <COMMA> AND ALIBABA <PERIOD>']
torch.Size([1, 274823, 80])
['HERE ARE JUST A FEW OF THE SEEMINGLY UNREAL <COMMA> ALBEIT <COMMA> ENTIRELY TRUE HIGHLIGHTS FROM HIS RESUME <PERIOD>', "HE'S A FORMER MTV BREAK DANCER IN TAIWAN <COMMA>", 'THE ONE TIME NATIONAL CHINESE KICKBOXING CHAMPION <COMMA>', 'THE FIRST AMERICAN TO HOLD THE GUINNESS WORLD RECORD IN TANGO <COMMA> A HORSEBACK ARCHER <COMMA> PRINCETON UNIVERSITY GUEST LECTURER <COMMA> AND ANGEL INVESTOR WHO HAS RACKED UP A SERIOUS STRING OF ENTREPRENEURIAL MAGA HITS <COMMA>', 'INCLUDING UBER <COMMA> FACEBOOK <COMMA> TWITTER <COMMA> AND ALIBABA <PERIOD>']
torch.Size([1, 274823, 80])

```

### CODE REVIEW

**Comment by danpovey** [comment]
Great-- thanks!

**Comment by pzelasko** [comment]
I think you'd do best to copy my code from snowfall for GigaSpeech; there's a number of details that need to be addressed before you can actually run training, e.g. splitting into smaller utterances, lazy-reading the CutSet, etc.

**Comment by danpovey** [comment]
OK.
In this case we actually won't be using the transcripts, as we'll be using it for unsupervised.  We don't really need the cut information either, at least I wasn't planning to use it.  I guess we can drop that somehow?

**Comment by danpovey** [comment]
Hm, maybe it's better that we do what Piotr said.  Would you mind figuring this part out, Liyong?
The tricky thing is that we are co-training with Librispeech, so the epochs will be determined by the size of the Librispeech training data.  We should probably use the "L" or "XL" subsets, which are larger than Librispeech, although of course it's OK to test with a smaller subset.  What i am thinking we can do about the data loaders, is just recreate the Gigaspeech dataloaders whenever they run out of data.  This doesn't even have to be exactly synchronized between the training processes.

**Comment by danpovey** [comment]
.. and it's OK with me to use the same cut information that we'd use if we were using the supervisions, although I don't plan to actually use the text.

**Comment by glynpu** [comment]
Make a copy from Piotr's implementation with following modification:
1.  change snowfall depencies with icefall
2. change some interface to make compatible to latest lhotse
3. Add an interface called "inexhaustible_train_dataloaders"
4. remove specaugment transforms as required

>recreate the Gigaspeech dataloaders whenever they run out of data

The added "inexhaustible_train_dataloaders" will do this.

example usage(It will never stop since the dataloader is inexhaustible):
```
import argparse                                                                 
from gigaspeech_datamodule import GigaSpeechAsrDataModule                       
                                                                                
def get_parser():                                                               
    parser = argparse.ArgumentParser(                                           
        formatter_class=argparse.ArgumentDefaultsHelpFormatter                  
    )                                                                           
    return parser                                                               
                                                                                
if __name__ == '__main__':                                                      
    parser = get_parser()                                                       
    GigaSpeechAsrDataModule.add_arguments(parser)                               
    args = parser.parse_args()                                                  
    gigaspeech = GigaSpeechAsrDataModule(args)                                  
    train_dl = gigaspeech.inexhaustible_train_dataloaders()                     
    for idx, batch in enumerate(train_dl):                                      
        print(batch["inputs"].shape)                                            
        print(len(batch["supervisions"]["text"]))                               
        print(batch["supervisions"]["text"][0:2])
```

**Comment by glynpu** [comment]
Now option conflicts between librispeech and gigaspeech (e.g. they both have --max-duration) are solved in a quite ugly way.

example usage:
```
python example_giga_dataloader.py --max-duration 10  giga --max-duration=150 --feature-dir="exp/data/"
```
The former --maxduration 10 is for librispeech; 
options after **_giga_** are for gigaspeech.
I believe there exits an elegant way for this problem
.

File: egs/librispeech/ASR/conformer_ctc/gigaspeech_datamodule.py [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+        test_loaders = []
+        if not is_list:
+            cuts = [cuts]
+
+        for cuts_test in cuts:
+            logging.debug("About to create test dataset")
+            test = K2SpeechRecognitionDataset(
+                cuts_test,
+                input_strategy=(
+                    OnTheFlyFeatures(Fbank(FbankConfig(num_mel_bins=80)), num_workers=8)
```
Comment (pzelasko) [inline egs/librispeech/ASR/conformer_ctc/gigaspeech_datamodule.py:337]: Note: replacing this with `KaldifeatFbank` in latest Lhotse may speed up on-the-fly feature extraction by 20~30% (on CPU).

File: egs/librispeech/ASR/prepare_gigaspeech.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+    sup: SupervisionSegment, oov_pattern=re.compile(r"<(SIL|MUSIC|NOISE|OTHER)>")
+) -> bool:
+    return oov_pattern.search(sup.text) is None
+
+
+def main():
+    args = get_parser().parse_args()
+    dataset_parts = [args.subset, "DEV", "TEST"]
+    if args.subset in ["L", "XL"]:
+        assert is_module_available("pyarrow"), (
```
Comment (pzelasko) [inline egs/librispeech/ASR/prepare_gigaspeech.py:143]: This doesn't look right to me; `pyarrow` is long gone from Lhotse, which version of the recipe is that? I don't see this code in https://github.com/k2-fsa/snowfall/blob/master/egs/gigaspeech/asr/simple_v1/prepare.py

File: egs/librispeech/ASR/prepare_gigaspeech.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    )
+
+    ctx_suffix = get_context_suffix(args)
+
+    print("Feature extraction:")
+    extractor = Fbank(FbankConfig(num_mel_bins=80))
+    with get_executor() as ex:  # Initialize the executor only once.
+        for partition, manifests in gigaspeech_manifests.items():
+            # For L and XL partition we are going to store the manifest using pyarrow.
+            cuts_path_ext = "jsonl.gz" if partition not in ["L", "XL"] else "arrow"
```
Comment (pzelasko) [inline egs/librispeech/ASR/prepare_gigaspeech.py:184]: This is also from an older version, please try to replace this with the recipe code on current master in snowfall.

BTW I originally skipped feature extraction for L and XL because it was too slow and took too much memory. But since then @csukuangfj added chunked feature extraction in kaldifeat, and I implemented CUDA extraction in Lhotse. You can safely compute features for everything with the new mechanisms. See the following method on CutSet with examples:

https://github.com/lhotse-speech/lhotse/blob/ce1437c8127d3545f7c45a5677c44bbb5b8d4f35/lhotse/cut.py#L3358-L3410

File: egs/librispeech/ASR/conformer_ctc/gigaspeech_datamodule.py [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+            type=str2bool,
+            default=False,
+            help='Should we use only 1000 utterances for dev (speeds up training)'
+        )
+
+    def validate_args(self):
+        if self.args.giga_subset in ['L', 'XL']:
+            assert (
+                self.args.giga_shuffle == False
+            ), "For GigaSpeech L/XL, you must use --shuffle 0 to avoid eagerly reading pyarrow manifests."
```
Comment (pzelasko) [inline egs/librispeech/ASR/conformer_ctc/gigaspeech_datamodule.py:184]: As mentioned in other comment -- this file also seems to come from an older version of the recipe.

File: egs/librispeech/ASR/prepare_gigaspeech.py (hunk 2 above)
Comment (glynpu) [reply egs/librispeech/ASR/prepare_gigaspeech.py:143]: Will modify it to the master one.

I directly copy the file from this link https://github.com/k2-fsa/snowfall/pull/230/files 
The website defaultly show "the pyarrow version", which exactly not the same with the one in the master branch.

Comment (glynpu) [reply egs/librispeech/ASR/prepare_gigaspeech.py:143]: If you search pyarrow from previous link, you would got the "pyarrow" version prepare.py .
https://github.com/k2-fsa/snowfall/pull/230/files#diff-10ff23c2be6cdc66d1461d8ff40b8acee607ab6c2e5d734c7ce03b65765289c3R143

I always thought github would show the latest version, but it seems not so.

Comment (pzelasko) [reply egs/librispeech/ASR/prepare_gigaspeech.py:143]: I think I made follow-up PRs where I synchronized the changes with Lhotse and also improved GigaSpeech recipe.

Comment (glynpu) [reply egs/librispeech/ASR/prepare_gigaspeech.py:143]: I know what happened; 
I was told that https://github.com/k2-fsa/snowfall/pull/230 is Piotr's implementation about gigaspeech.

But some files related are also modified in pr https://github.com/k2-fsa/snowfall/pull/238

That's why the older version is refered.

File: egs/librispeech/ASR/prepare_gigaspeech.py (hunk 3 above)
Comment (glynpu) [reply egs/librispeech/ASR/prepare_gigaspeech.py:184]: >took too much memory

How is this problem solved?
I think originally feats are already stored in HDF5 format.

Comment (glynpu) [reply egs/librispeech/ASR/prepare_gigaspeech.py:184]: For gigaspeech, do you have any recommand values of num_workers and batch_duration when using compute_and_store_features_batch?
Now the default values are used, i.e. num_workers=4 and batch_duration=600.0

File: egs/librispeech/ASR/conformer_ctc/gigaspeech_datamodule.py (hunk 4 above)
Comment (glynpu) [reply egs/librispeech/ASR/conformer_ctc/gigaspeech_datamodule.py:212]: This part is the same with snowfall master.  https://github.com/k2-fsa/snowfall/blob/master/egs/gigaspeech/asr/simple_v1/asr_datamodule.py#L183

Shall I removed the check of shuffle==False and bucketing_sampler==False? (I already removed the check of on_the_fly_feats)

File: egs/librispeech/ASR/prepare_gigaspeech.py (hunk 3 above)
Comment (pzelasko) [reply egs/librispeech/ASR/prepare_gigaspeech.py]: > > took too much memory
> 
> How is this problem solved? I think originally feats are already stored in HDF5 format.

It’s about the RAM memory, not disk usage. Basically for 1h+ waveforms naive feature extraction implementations use quadratic memory. Kaldifeat has a mechanism to prevent that. But anyway without CUDA it would have taken too much time to compute feats for 10k hours.

> For gigaspeech, do you have any recommand values of num_workers and batch_duration when using compute_and_store_features_batch?
Now the default values are used, i.e. num_workers=4 and batch_duration=600.0

It should work fine with defaults; you can try increasing these but I found that on some systems it generates too much I/O if you create too many workers and the DataLoader crashes.

File: egs/librispeech/ASR/conformer_ctc/gigaspeech_datamodule.py (hunk 4 above)
Comment (pzelasko) [reply egs/librispeech/ASR/conformer_ctc/gigaspeech_datamodule.py:212]: You can remove the shuffle==False check (I added streaming pseudo shuffling in Lhotse) but I’m not sure about the BucketingSampler — I think I still haven’t added support for very large data for it (it will take long time, a few minutes maybe, to start the script if you turn it on). I can try to add it soon.
